### PR TITLE
Bulk annotations right panel

### PR DESCRIPTION
**Pull request by will-moore**
This is metadata "Bulk Annotations" PR https://github.com/openmicroscopy/openmicroscopy/pull/3875 rebased onto develop (actually it was on develop originally as https://github.com/openmicroscopy/openmicroscopy/pull/3863 but not merged).

Please see #3875 for main testing info.

NB: Also tweaked to add this under "Tables" collapsible tab. To test:
 - When "Tables" tab is not expanded, the data is not loaded.
 - When you expand the "Tables" tab, data will be loaded
 - If the "Tables" tab is expanded, selecting new Well will load new data.
 - "Info" should give tooltip (currently not working - known issue)

screen shot 2016-02-04 at 15 26 32

### CODE REVIEW

**Comment by will-moore** [comment]
Do we need to update our Docs or Help guides with this functionality, since this is otherwise a "hidden feature"? cc @hflynn @gusferguson. Only described in our testing scenario at https://github.com/openmicroscopy/ome-internal/blob/master/testing_scenarios/BulkAnnotations.txt

**Comment by joshmoore** [comment]
The ImageJ folks were certainly asking recently about "how do I upload a table from ImageJ to have it shown in the right-hand panel".

**Comment by gusferguson** [comment]
Will this ever be "User" functionality or will it really be an Admin/Facility Manager/Developer functionality?

**Comment by joshmoore** [comment]
Probably not a completely vanilla user concept until some file formats come with pre-defined bulk annotations that the users will want to see in the UI. For the moment, perhaps more power-users who are willing to add a file in the proper format to add this value.

**Comment by jburel** [comment]
@gusferguson: it will probably be like roi-folder first use cases.
Table generated via a script and viewable in the clients

File: components/tools/OmeroWeb/omeroweb/webgateway/views.py [hunk 1]
```diff
@@ -2267,24 +2267,42 @@ def _annotations(request, objtype, objid, conn=None, **kwargs):
     query += """
```
Comment (emilroz) [inline components/tools/OmeroWeb/omeroweb/webgateway/views.py]: @will-moore without extra ```join fetch links.details.owner``` this query will not load owner details when working in none default groups, which results in exception later on in the code.

**Comment by will-moore** [comment]
OK, so we probably need some code examples in Python, Java and Matlab.

**Comment by jburel** [comment]
@joshmoore: how large are the "real" tables that you have seen during work on IDR?
some performance testing probably required here since loading the full table could become problematic.

**Comment by joshmoore** [comment]
Largest file size is at the moment is 740,340,765 (as returned by HQL). The `ptdump` info on that file is:

```
/OME/Measurements (Table(370176,)) ''
  description := {
  "Plate": Int64Col(shape=(), dflt=0, pos=0),
  "Plate Group": StringCol(itemsize=3, shape=(), dflt='', pos=1),
  "Well Number": StringCol(itemsize=3, shape=(), dflt='', pos=2),
  "Well": Int64Col(shape=(), dflt=0, pos=3),
```

i.e. 370K rows. The data columns are:

```
  "siRNA Identifier": StringCol(itemsize=6, shape=(), dflt='', pos=4),
  "Antisense sequence": StringCol(itemsize=21, shape=(), dflt='', pos=5),
  "Sense Sequence": StringCol(itemsize=21, shape=(), dflt='', pos=6),
  "Gene Identifier": StringCol(itemsize=383, shape=(), dflt='', pos=7),
  "Gene Symbol (Ensembl 81)": StringCol(itemsize=15, shape=(), dflt='', pos=8),
  "Control Type": StringCol(itemsize=16, shape=(), dflt='', pos=9),
  "Control Comments": StringCol(itemsize=19, shape=(), dflt='', pos=10),
  "Quality Control": StringCol(itemsize=4, shape=(), dflt='', pos=11),
  "Comments": StringCol(itemsize=51, shape=(), dflt='', pos=12),
  "Plate Issues": StringCol(itemsize=1, shape=(), dflt='', pos=13),
  "Gene Symbol (Ensembl 61)": StringCol(itemsize=21, shape=(), dflt='', pos=14),
  "Deviation": StringCol(itemsize=6, shape=(), dflt='', pos=15),
  "Cell Number": StringCol(itemsize=3, shape=(), dflt='', pos=16),
  "Valid Plates": StringCol(itemsize=1, shape=(), dflt='', pos=17),
  "% Inhibitor": StringCol(itemsize=4, shape=(), dflt='', pos=18),
  "% Accelerator": StringCol(itemsize=4, shape=(), dflt='', pos=19),
  "Transcripts Hit": StringCol(itemsize=5, shape=(), dflt='', pos=20),
  "Transcript Identifiers": StringCol(itemsize=1135, shape=(), dflt='', pos=21),
  "Gene In Validation Screen": StringCol(itemsize=3, shape=(), dflt='', pos=22),
  "Strong Inhibitor Gene Hit": StringCol(itemsize=3, shape=(), dflt='', pos=23),
  "Phenotype 1": StringCol(itemsize=18, shape=(), dflt='', pos=24),
  "Phenotype 1 Term Name": StringCol(itemsize=35, shape=(), dflt='', pos=25),
  "Phenotype 1 Term Accession": StringCol(itemsize=12, shape=(), dflt='', pos=26),
  "Phenotype 2": StringCol(itemsize=28, shape=(), dflt='', pos=27),
  "Phenotype 2 Term Name": StringCol(itemsize=42, shape=(), dflt='', pos=28),
  "Phenotype 2 Term Accession": StringCol(itemsize=12, shape=(), dflt='', pos=29),
  "Phenotype 3": StringCol(itemsize=30, shape=(), dflt='', pos=30),
  "Phenotype 3 Term Name": StringCol(itemsize=44, shape=(), dflt='', pos=31),
  "Phenotype 3 Term Accession": StringCol(itemsize=12, shape=(), dflt='', pos=32),
  "Plate Name": StringCol(itemsize=19, shape=(), dflt='', pos=33),
  "Well Name": StringCol(itemsize=3, shape=(), dflt='', pos=34)}
  byteorder := 'little'
  chunkshape := (65,)
```

**Comment by will-moore** [comment]
Robot tests failing https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/213/console with:
```
08:54:11 + python lib/python/omero/util/populate_metadata.py -k 95d21029-02eb-4fcd-a794-6a7f2689f8d1 Plate:282 bulk_annotation.csv
08:54:12 Traceback (most recent call last):
08:54:12   File "lib/python/omero/util/populate_metadata.py", line 533, in <module>
08:54:12     client.joinSession(session_key)
08:54:12   File "/opt/hudson/workspace/OMERO-DEV-merge-robotframework/OMERO.server-5.2.1-348-2b428d1-ice35-b221/lib/python/omero/clients.py", line 496, in joinSession
...
08:54:12 Ice.ConnectionRefusedException: Ice.ConnectionRefusedException:
08:54:12 Connection refused
```

**Comment by will-moore** [comment]
Although this seems to work fine locally:
```
+ python lib/python/omero/util/populate_metadata.py -k 93ec89e8-7e50-4dce-8d2d-4a6f5d7715f7 Plate:3167 bulk_annotation.csv
INFO:omero.util.Resources:Starting
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Created new table OriginalFile:42999
INFO:omero.util.populate_metadata:Table initialized with 4 columns.
INFO:omero.util.populate_metadata:Added data column data.
INFO:omero.util.Resources:Halted
```

File: components/tools/OmeroWeb/omeroweb/webgateway/views.py (hunk 1 above)
Comment (will-moore) [inline components/tools/OmeroWeb/omeroweb/webgateway/views.py]: Thanks @emilroz, but it seems to be loading link owners already.
In the json at webgateway/table/Plate.wells/2061/query/?query=Well-2061 I see 
``` "addedBy": "Will Moore" ```
even when logged in as another user whose default group is different from this one.

Comment (emilroz) [inline components/tools/OmeroWeb/omeroweb/webgateway/views.py]: OK, that wasn't my experience at all. I could only load this is the user's default group was set to the group where the data was. @will-moore are you testing with 5.2?

Comment (will-moore) [inline components/tools/OmeroWeb/omeroweb/webgateway/views.py]: Yes, on develop branch.

Comment (emilroz) [inline components/tools/OmeroWeb/omeroweb/webgateway/views.py]: All the test I've done where on 5.1 so maybe there's a difference to how the objects are loaded now.

Comment (will-moore) [inline components/tools/OmeroWeb/omeroweb/webgateway/views.py]: @emilroz I went to add a test for loading of table data for plate. Although I didn't test the permissions in the test itself, I found that for ```root``` user to query data from another user in a private group *does* need to fetch ```links.details.owner```. Thanks!

**Comment by will-moore** [comment]
@sbesson Any idea why I'm getting ```Ice.ConnectionRefusedException``` in the robot setup script above  (when it works locally) with
```
python lib/python/omero/util/populate_metadata.py -k 95d21029-02eb-4fcd-a794-6a7f2689f8d1 Plate:282 bulk_annotation.csv
```

**Comment by sbesson** [comment]
@will-moore: have you tried passing the server_name and port number to the script? In the CI job, the script is executed against a remote server and this error would make sense if the script was trying to connect to `localhost` with the session key.

File: components/tests/ui/robot_setup.sh [hunk 2]
```diff
@@ -28,13 +29,21 @@ bin/omero user add $USER_NAME $USER_NAME $USER_NAME $GROUP_NAME $GROUP_NAME_2 --
...
 
+# Create batch annotation csv
+echo "Well,Well Type,Concentration" > "$BULK_ANNOTATION_CSV"
+echo "A1,Control,0" >> "$BULK_ANNOTATION_CSV"
+echo "A2,Treatment,10" >> "$BULK_ANNOTATION_CSV"
+
 # Create robot setup
 bin/omero login $USER_NAME@$HOSTNAME:$PORT -w $USER_PASSWORD
+# Parse the sessions file to get session key
+key=$(grep omero.sess $(bin/omero sessions file) | cut -d= -f2)
```
Comment (sbesson) [inline components/tests/ui/robot_setup.sh]: You should be able to use `bin/omero session key` directly

**Comment by pwalczysko** [comment]
Tested today, see testing remarks on https://github.com/ome/scripts/pull/107#issuecomment-180320777.

**Comment by pwalczysko** [comment]
Note that the known bug, which is demonstrating as doubling of the output under ``Tables`` tab comes up when the full viewer of a next well is opened using double-click onto thumbnail. Interestingly, the double-click must be fast, if there is a pause between the clicks of the double-click, then the bug does not appear.

**Comment by will-moore** [comment]
@pwalczysko That commit will fix the duplication in right panel (if not the underlying cause).

**Comment by will-moore** [comment]
Also need to check that Image Viewer shows the same data (under "Image Information")

**Comment by hflynn** [comment]
@gusferguson what is your thinking here? Have you made a Trello card for adding this functionality to a help guide?

**Comment by pwalczysko** [comment]
@will-moore : Yes, the https://github.com/openmicroscopy/openmicroscopy/pull/4446#issuecomment-181332716 duplication problem is gone. The image Viewer info https://github.com/openmicroscopy/openmicroscopy/pull/4446#issuecomment-181333231 check I have done previously and confirmed today that the info is displayed and okey.

**Comment by will-moore** [comment]
Discussed with @pwalczysko: The webclient functionality & tests are working OK.
We have other PRs for Python code samples and an OMERO.script for populating metadata. We just need to have some 'help' docs on how to use this cc @gusferguson.

The only outstanding question from above is @jburel asking about performance, of "loading the full table". I don't know what our criteria should be for performance: we maybe want to test with a moderately large file, but I don't think a table with 370K rows is realistic since this scenario is one row per well, E.g. 364 max. Although we do ```openTable(file)``` to open the file, we only query a row at a time for display in webclient.

@joshmoore Should I generate a large csv file (how large) and test loading speed? If so, how slow is too slow? Or, are we happy that the strategy here is the optimum way of reading tables? NB: this PR doesn't change any of the tables queries since they were already in use for the image viewer (and have been in use on the IDR for some time).

**Comment by emilroz** [comment]
@will-moore one thing to keep in mind here, is that the table can be attached also to the screen. So for example: http://jcb-dataviewer.rupress.org/jcb/browse/7555/ you could have a table with: 141 plates, 3 fields, 96 wells: 41K rows. We have examples where number of fields is more then 10 with 384 wells so we can easily talk here about 100's of thousands of rows per table.

**Comment by will-moore** [comment]
@emilroz This PR doesn't support use-cases with multiple Fields (rows) per Well, and I'm not sure if (and populate_metadata.py does? But I can see that it's possible to support multiple Plates per Screen.

**Comment by will-moore** [comment]
NB: I noticed that Tables on Screens breaks when the Plate is in multiple Screens.
```
    serverStackTrace = ome.conditions.ApiUsageException: Query named:
	select obj0 from Screen obj0
...
has returned more than one Object
findBy methods must return a single value.
```
For now, I will fix this to use findBy query, and pick the first Screen. 
But this comes back to the question discussed with @pwalczysko above about improving the display of multiple tables for each well. But that should be handled in a future PR.

**Comment by will-moore** [comment]
To test the last commit, follow the Bulk Annotation scenario for a table on a Screen, then put the Plate into a second Screen and check that the tables data is still displayed OK on Wells.

**Comment by emilroz** [comment]
@will-moore we have added support for Images in this commit: https://github.com/joshmoore/openmicroscopy/commit/29b9ddd139b42d7b584252980a52635604fdeb1d
so it is possible to resolve the metadata on Images instead of wells. I have the changes ready so I can add the commits that introduce the Image queries as well.

**Comment by pwalczysko** [comment]
The test of the last commit passed https://github.com/openmicroscopy/openmicroscopy/pull/4446#issuecomment-183091657 - I suppose some further discussion should happen between @will-moore and @emilroz considering the steps here.

**Comment by will-moore** [comment]
Thanks @emilroz. I am away Monday & Tuesday, and it would be nice to get this PR in, so I can begin on rebasing #3896. Could we add those commits into a different PR?

**Comment by will-moore** [comment]
@joshmoore Is there anything else needed for this PR? See questions from https://github.com/openmicroscopy/openmicroscopy/pull/4446#issuecomment-182790997
Otherwise, it would be great to get this merged so that I can start rebasing #3896 on top of it.

**Comment by joshmoore** [comment]
> @joshmoore Should I generate a large csv file (how large) and test loading speed? If so, how slow is too slow?

Sorry, I missed the direct question on me, @will-moore. From my side, this is currently working in "production" but I leave suggestions for pounding it further on the mainline to others if they have them.

I think probably more necessary is a final code review. @chris-allan had pointed out the change in `marshal.py` as a possible concern. Perhaps you could go through and explain any changes that might need to be changed in the doc, @will-moore? Just need to make sure the impact for a point release is within bounds.

**Comment by will-moore** [comment]
I am happy to revert the change to marshal.py. Agreed, it is an API change, although it could be considered a bug fix! I just found that because of a script I had an image in a Well and a Dataset (E.g. you can get this with Dataset_To_Plate script) which broke the display of bulk annotations in the viewer since imgData json only handles a single Parent.

**Comment by will-moore** [comment]
@joshmoore @chris-allan Reverted the changes to ```marshal.py```.

**Comment by chris-allan** [comment]
For what it's worth we have the exact same problems at the moment with the PathViewer. If images are in multiple datasets, as the image data JSON only currently supports a single parent, things quickly go off the rails.

What I'd suggest is a semi-additive change where the first Dataset and WellSample are included in the existing fields, rather than the last as in 0c02e69ec784b9140e5ce92a5ecb7e58dd8759f3, and we include a list of all parents under the `parents` or similar key. This is "breaking" but only as far as semantics are concerned and is in line with the intentions of the existing JSON. What we should examine quite closely though is the use of these keys in the full image viewer before making any change at all.

/cc @knabar

**Comment by will-moore** [comment]
@chris-allan I decided to hand parents in ```imageMarshal``` in a different PR, so as not to hold up this one https://github.com/openmicroscopy/openmicroscopy/pull/4486

**Comment by jburel** [comment]
"API" PR not included in this release, use cases/implementations to be discussed.
Merging this PR

**Pull request merged**